### ISSUE TITLE
Intel C++ compiler 17 fails to compile overload_cast based on constness

### ISSUE THREAD

**Issue by Baljak** [pybind/pybind11#1607]
This is the case at least on Windows, using Intel C++ compiler 17.0. This is fixed starting ICC 18.0. Since this is a compiler related bug, I'm not sure it can be fixed. At least, ICC version required for Windows could be set to 18.0 or newer.

```c++
#include <pybind11/pybind11.h>

namespace py = pybind11;

class foo
{
    int bar_;
public:
    explicit foo(int bar) :bar_(bar) {}

    int& bar() { return bar_; }
    int bar() const { return bar_; }
};

PYBIND11_MODULE(test, m) 
{
    py::class_<foo>(m, "foo")
        .def("bar", py::overload_cast<>(&foo::bar, py::const_));
}
```

Error raised by ICC 17.0:
```
C:\Dev\tests\pybind11\module.cpp(18): error : no instance of overloaded function "pybind11::detail::overload_cast_impl<Args...>::operator() [with Args=<>]" matches the argument list
1>              argument types are: (<unknown-type>, const std::integral_constant<bool, 1>)
1>              object type is: const pybind11::detail::overload_cast_impl<>
1>            .def("bar", py::overload_cast<>(&foo::bar, py::const_));
1>                        ^
1>  C:\Dev\tests\pybind11\pybind11-2.2.4\include\pybind11/detail/common.h(736): note: this candidate was rejected because at least one template argument could not be deduced
1>        constexpr auto operator()(Return (Class::*pmf)(Args...) const, std::true_type) const noexcept
1>                       ^
1>  C:\Dev\tests\pybind11\pybind11-2.2.4\include\pybind11/detail/common.h(732): note: this candidate was rejected because arguments do not match
1>        constexpr auto operator()(Return (Class::*pmf)(Args...), std::false_type = {}) const noexcept
1>                       ^
1>  C:\Dev\tests\pybind11\pybind11-2.2.4\include\pybind11/detail/common.h(728): note: this candidate was rejected because mismatch in count of arguments
1>        constexpr auto operator()(Return (*pf)(Args...)) const noexcept
1>                       ^
1>
1>  compilation aborted for C:\Dev\tests\pybind11\module.cpp (code 2)
```

**Comment by wjakob** [comment]
Intel really ought to test a bit more against pybind11. There is no end of bugs from this compiler -- every new release breaks in some different way :(

**Comment by Baljak** [comment]
I agree, I managed to compile my project with ICC 18.0 however recently released 19.0 brings new segmentation faults problems...

**Comment by YannickJadoul** [comment]
The original issue seems to be fixed, so I'll close this. For more ICC errors, please open a new issue :-)
